### ISSUE TITLE
problem with \file attribute and soft-link

### ISSUE THREAD

**Issue by aotto1968** [doxygen/doxygen#9413]
**Describe the bug**
I get the following error:
`…/tcl_MpGrep.doc:2: warning: the name '…/MpGrep.doc' supplied as the argument in the \file statement is not an input file`

The problem is that `tcl_MpGrep.doc` is a soft-link from `MpGrep.doc` and `MpGrep.doc` has the header

```
/**
\file       theLink/docs/man/MpGrep.doc
\brief      \$Id: pages.doc 341 2020-01-22 21:32:38Z dev1usr $
…
```

> ls -al tcl_MpGrep.doc
lrwxrwxrwx 1 dev1usr users 15 26. Feb 2021  tcl_MpGrep.doc -> MpGrep.doc

the `MpGrep.doc` is used as template for different files (tcl,cc,…), so that I can't just rename it.

**Expected behavior**
I expect that that the `\file` command is able to deal with _soft-link_

**To Reproduce**
just do a simple _soft-link_ from a **.doc** file.

**Version**
1.9.3

**Comment by albert-github** [comment]
Recently some improvements have been made in respect to links, which landed in the current master.
- Please have a try with the current master (1.9.5 (886349348ad5dc60152e0d5b92e6ffbaf5f4a921)
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that shows the problem that should be fixed with the proposed patch? Please don't add external links as they might not be persistent.

**Comment by albert-github** [comment]
A quick revisit and attempt with the current master, and this looks OK to me.
The example setup is like:
```
.:
total 17
 4 drwxrwxr-x+ 1 Auser  Agrp  0 Jun 20 17:06 .
 1 -rw-rw-r--+ 1 Auser  Agrp 36 Jun 20 17:04 Doxyfile
 0 lrwxrwxrwx  1 Auser  Agrp 20 Jun 20 17:03 the_real_file.h -> orig/the_real_file.h
 0 drwxrwxr-x+ 1 Auser  Agrp  0 Jun 20 17:02 orig
12 drwxrwx---+ 1 Auser  Agrp  0 Jun 20 17:00 ..

./orig:
total 5
4 drwxrwxr-x+ 1 Auser  Agrp  0 Jun 20 17:06 ..
0 drwxrwxr-x+ 1 Auser  Agrp  0 Jun 20 17:02 .
1 -rw-rw-r--+ 1 Auser  Agrp 45 Jun 20 17:01 the_real_file.h
```

With **Doxyfile**
```
QUIET = YES
INPUT = the_real_file.h
```
and content of the **orig/the_real_file.h**
```
/// \file
///


/// the function
void fie();
```

The result is (file list page):
![image](https://user-images.githubusercontent.com/5223533/174633748-2a9f45d3-cdd6-4c1b-868b-cb42e3f5031b.png)


Note the fact that I didn't use a filename with the `\file` command, it is a common misconception that it is mandatory to have a filename. It is only necessary when the documentation of a file is in another file that the file where it belongs to.

Does the given setup solve your problem?

Example project: [example.tar.gz](https://github.com/doxygen/doxygen/files/8941884/example.tar.gz)

**Comment by aotto1968** [comment]
Try to build the latest "doxygen" but the build failed.

```
[  1%] Built target md5
[  1%] Built target lodepng
[ 10%] Built target mscgen
[ 10%] Checking the git repository for changes...
[ 10%] Built target check_git_repository
[ 10%] Checking the doxygen version for changes...
[ 10%] Built target check_doxygen_version
[ 11%] Built target doxygen_version
[ 13%] Built target xml
[ 14%] Built target generate_configvalues_header
[ 17%] Built target vhdlparser
make[2]: Circular generated_src/configimpl.corr <- src/configimpl.l dependency dropped.
[ 17%] Generating configimpl.l
[ 18%] [FLEX][configimpl] Building scanner with flex 2.6.4
[ 18%] Generating ../generated_src/configimpl.cpp
[ 18%] Generating ../generated_src/configimpl.l.h
Consolidate compiler generated dependencies of target doxycfg
make[2]: Circular generated_src/configimpl.l <- generated_src/configimpl.corr dependency dropped.
[ 18%] Generating ../generated_src/configimpl.corr, ../generated_src/configimpl.l
[ 18%] Generating ../generated_src/configimpl.cpp
[ 19%] Building CXX object src/CMakeFiles/doxycfg.dir/__/generated_src/configimpl.cpp.o
[ 19%] Linking CXX static library ../lib/libdoxycfg.a
[ 23%] Built target doxycfg
make[2]: Circular generated_src/xmlcode.corr <- src/xmlcode.l dependency dropped.
make[2]: Circular src/constexp.y <- src/constexp.y dependency dropped.
make[2]: Circular generated_src/code.l <- generated_src/code.corr dependency dropped.
[ 24%] Generating code.l
Error copying file "…/test/doxygen/generated_src/code.l" to "…/test/doxygen/src/code.l".
make[2]: *** [src/CMakeFiles/doxymain.dir/build.make:643: src/code.l] Error 1
make[1]: *** [CMakeFiles/Makefile2:511: src/CMakeFiles/doxymain.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

**Comment by albert-github** [comment]
Strange those build errors, especially the
```
Error copying file "…/test/doxygen/generated_src/code.l" to "…/test/doxygen/src/code.l".
```
doxygen doesn't copy the file code.l from generated_src to src.
Also the Circular dependency is strange.
I have no problems building.
- What were the commands you used?
- did you build in place (i.e. in the doxygen root) or out of place (e.g. in a subdrirectory ´build`, it s possible that the in place version gives some problems (I'll have to investigate this, will take some time) and is, in my opinion, also bad practice especially for usage when building for multiple platforms or multiple setups with CMake.

- Why use the filename with the `\file`?
- I don*t think the directory name makes a difference

**Comment by aotto1968** [comment]
Hi, I using the git directory to build:

```
:~/tmp/test/doxygen> ls -al
insgesamt 136
drwxr-xr-x 24 dev1usr users   820 20. Jun 20:58 .
drwxr-xr-x  6 dev1usr users   160 20. Jun 20:55 ..
drwxr-xr-x  9 dev1usr users   240 20. Jun 20:58 addon
drwxr-xr-x  2 dev1usr users    40 20. Jun 20:58 bin
-rw-r--r--  1 dev1usr users  2287 20. Jun 20:55 BUILD.txt
drwxr-xr-x  3 dev1usr users   320 20. Jun 20:55 cmake
-rw-r--r--  1 dev1usr users 21193 20. Jun 20:58 CMakeCache.txt
drwxr-xr-x  5 dev1usr users   320 20. Jun 21:03 CMakeFiles
-rw-r--r--  1 dev1usr users  2378 20. Jun 20:58 cmake_install.cmake
-rw-r--r--  1 dev1usr users  7805 20. Jun 20:55 CMakeLists.txt
-rw-r--r--  1 dev1usr users   170 20. Jun 20:55 .codedocs
-rw-r--r--  1 dev1usr users  7609 20. Jun 20:58 CPackConfig.cmake
-rw-r--r--  1 dev1usr users  8037 20. Jun 20:58 CPackSourceConfig.cmake
-rw-r--r--  1 dev1usr users   447 20. Jun 20:58 CTestTestfile.cmake
drwxr-xr-x  2 dev1usr users  1740 20. Jun 20:55 doc
-rw-r--r--  1 dev1usr users 11356 20. Jun 20:55 Doxyfile
-rw-r--r--  1 dev1usr users    96 20. Jun 20:55 .editorconfig
drwxr-xr-x  2 dev1usr users  1320 20. Jun 20:55 examples
drwxr-xr-x  2 dev1usr users    60 20. Jun 20:55 filesystem
drwxr-xr-x  2 dev1usr users   580 20. Jun 21:00 generated_src
drwxr-xr-x  8 dev1usr users   260 20. Jun 21:03 .git
drwxr-xr-x  4 dev1usr users   100 20. Jun 20:55 .github
-rw-r--r--  1 dev1usr users   159 20. Jun 20:55 .gitignore
-rw-r--r--  1 dev1usr users   152 20. Jun 20:55 INSTALL
drwxr-xr-x  3 dev1usr users   280 20. Jun 20:55 jquery
-rw-r--r--  1 dev1usr users  2269 20. Jun 20:55 LANGUAGE.HOWTO
drwxr-xr-x  2 dev1usr users   180 20. Jun 21:03 lib
drwxr-xr-x  3 dev1usr users   160 20. Jun 20:58 liblodepng
drwxr-xr-x  3 dev1usr users   180 20. Jun 20:58 libmd5
drwxr-xr-x  3 dev1usr users   920 20. Jun 20:58 libmscgen
drwxr-xr-x  3 dev1usr users   200 20. Jun 20:58 libversion
drwxr-xr-x  3 dev1usr users   160 20. Jun 20:58 libxml
-rw-r--r--  1 dev1usr users 18092 20. Jun 20:55 LICENSE
-rw-r--r--  1 dev1usr users 13446 20. Jun 20:58 Makefile
-rw-r--r--  1 dev1usr users  3099 20. Jun 20:55 README.md
drwxr-xr-x  3 dev1usr users  6380 20. Jun 21:03 src
drwxr-xr-x  6 dev1usr users   120 20. Jun 20:55 templates
drwxr-xr-x 86 dev1usr users  3660 20. Jun 20:58 testing
-rw-r--r--  1 dev1usr users     6 20. Jun 20:55 VERSION
drwxr-xr-x  3 dev1usr users   540 20. Jun 20:58 vhdlparser
drwxr-xr-x  2 dev1usr users   120 20. Jun 20:55 winbuild

:~/tmp/test/doxygen> LANG=C make
[  1%] Built target md5
[  1%] Built target lodepng
[ 10%] Built target mscgen
[ 10%] Checking the git repository for changes...
[ 10%] Built target check_git_repository
[ 10%] Checking the doxygen version for changes...
[ 10%] Built target check_doxygen_version
[ 11%] Built target doxygen_version
[ 13%] Built target xml
[ 14%] Built target generate_configvalues_header
[ 17%] Built target vhdlparser
make[2]: Circular generated_src/configimpl.corr <- src/configimpl.l dependency dropped.
[ 17%] Generating configimpl.l
[ 18%] [FLEX][configimpl] Building scanner with flex 2.6.4
[ 18%] Generating ../generated_src/configimpl.cpp
[ 18%] Generating ../generated_src/configimpl.l.h
Consolidate compiler generated dependencies of target doxycfg
make[2]: Circular generated_src/configimpl.l <- generated_src/configimpl.corr dependency dropped.
[ 18%] Generating ../generated_src/configimpl.corr, ../generated_src/configimpl.l
[ 18%] Generating ../generated_src/configimpl.cpp
[ 19%] Building CXX object src/CMakeFiles/doxycfg.dir/__/generated_src/configimpl.cpp.o
[ 19%] Linking CXX static library ../lib/libdoxycfg.a
[ 23%] Built target doxycfg
make[2]: Circular generated_src/xmlcode.corr <- src/xmlcode.l dependency dropped.
make[2]: Circular src/constexp.y <- src/constexp.y dependency dropped.
make[2]: Circular generated_src/code.l <- generated_src/code.corr dependency dropped.
[ 24%] Generating code.l
Error copying file "/home/dev1usr/tmp/test/doxygen/generated_src/code.l" to "/home/dev1usr/tmp/test/doxygen/src/code.l".
make[2]: *** [src/CMakeFiles/doxymain.dir/build.make:643: src/code.l] Error 1
make[1]: *** [CMakeFiles/Makefile2:511: src/CMakeFiles/doxymain.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

if I create a "build" subdirectory than the Makefile is written to the source-directory ???

```
dev1usr@linux02:~/tmp/test/doxygen> cd build/
dev1usr@linux02:~/tmp/test/doxygen/build> cmake -G "Unix Makefiles" ..
-- The javacc executable: /usr/bin/javacc (found version "7.0.11")
-- Configuring done
-- Generating done
-- Build files have been written to: /home/dev1usr/tmp/test/doxygen
dev1usr@linux02:~/tmp/test/doxygen/build> ls -al
insgesamt 0
drwxr-xr-x  2 dev1usr users  40 21. Jun 09:25 .
drwxr-xr-x 25 dev1usr users 840 21. Jun 09:26 ..
```

**Comment by albert-github** [comment]
The procedure you describe like;
- git to the git directory
- mkdir build
- cd build
- cmake -G "Unix Makefiles ..
- make 

looks OK, A bit strange is the short output of the cmake command and the fact that after the cmake part the build directory still looks / is empty. Also the reported fact that there is a Makefile created in the git directory is strange
(By the way: which version of cmake are you using?)

**Comment by aotto1968** [comment]
There is no Makefile in the "build" subdirectory of "test"

```
dev1usr@linux02:~/tmp/test/doxygen/build> cmake -G "Unix Makefiles" .. 
-- The javacc executable: /usr/bin/javacc (found version "7.0.11")
-- Configuring done
-- Generating done
-- Build files have been written to: /home/dev1usr/tmp/test/doxygen
dev1usr@linux02:~/tmp/test/doxygen/build> ls -al
total 0
drwxr-xr-x  2 dev1usr users  40 Jun 21 09:25 .
drwxr-xr-x 25 dev1usr users 840 Jun 22 08:18 ..
dev1usr@linux02:~/tmp/test/doxygen/build> make
make: *** No targets specified and no makefile found.  Stop.
```

**Comment by albert-github** [comment]
Probably the lines:
```
dev1usr@linux02:~/tmp/test/doxygen/build> cmake -G "Unix Makefiles" .. 
CMake Warning:
  Ignoring extra path from command line:

   ".."


CMake Error: The source directory "/home/dev1usr/tmp/test/doxygen" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

give the clue to the problem.

Do you have an alias or script defined for the cmake command that automatically adds a `.`as path?
In that case cmake can take the first as path and give a warning for the second.

**Comment by aotto1968** [comment]
```
dev1usr@linux02:~/tmp/test/doxygen/build> which cmake
/usr/bin/cmake
dev1usr@linux02:~/tmp/test/doxygen/build> rpm -qif /usr/bin/cmake
Name        : cmake-full
Version     : 3.23.2
Release     : 1.1
Architecture: x86_64
Install Date: Mo 20 Jun 2022 20:57:44 CEST
Group       : Unspecified
Size        : 34834292
License     : BSD-3-Clause
Signature   : RSA/SHA256, Mo 06 Jun 2022 11:44:19 CEST, Key ID b88b2fd43dbdc284
Source RPM  : cmake-full-3.23.2-1.1.src.rpm
Build Date  : Mo 06 Jun 2022 11:17:45 CEST
Build Host  : lamb23
Packager    : https://bugs.opensuse.org
Vendor      : openSUSE
URL         : https://www.cmake.org/
Summary     : Cross-platform make system
Description :
CMake is a cross-platform build system.
Distribution: openSUSE Tumbleweed
```

**Comment by aotto1968** [comment]
Ok, the **build** problem seems to be solved
→ there was a collection of some problems.
1. first I have done a build **in** the _doxygen_ directory → this seems to be the **first** problem.
2. than I make s subdirectory _build_ but this has a problem that a **previous** build in the _doxygen_ directory produces some *extra* files which **missconfigure** the build in the _build_ directory, this was the **second** problem
3. after I delete the entire _doxygen_ directory to start from scratch BUT create the _doxygen_ directory first. `git clone…`
create the _doxygen/doxygen_ directory and my _build_ directory in the **first** _doxygen_ directory does **not** find the source, this was the **third** problem.

**Comment by albert-github** [comment]
Looks like some good progress.

The problem you have now, as far as I can see, is that another tool already writes the header with the `\file orig/the_real_file.tmpl`, so we have to look for a way to adjust the specific line. 
So actually not really a doxygen problem but the problem of the application writing the header-

I think can doxygen can be of help here though by means of creating a filter where the line like:
```
/// \file the_real_file.tmpl
```
by the line:
```
/// \file
```
one could place e.g. the following line in the doxygen settings file:
```
FILTER_PATTERNS = *.h="sed -e \"s/\\\\file .*/\\\\file/\""
```

Best is to read up on the filter possibilities in doxygen.

**Comment by albert-github** [comment]
Regarding the in-place build I've just pushed a proposed patch, pull request #9423

**Comment by albert-github** [comment]
Regarding the in-place build the code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

The problem regarding the name of the file is in my opinion usage and can be solved by means of:
- a filter in doxygen
- the program creating the file doesn't write the file name in the `\file` statement.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.5.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
